### PR TITLE
Refactor Expr::BinaryExpr to use a struct

### PR DESCRIPTION
**Pull request by zhoudongyan**
# Which issue does this PR close?

Part of #2175 and #3807

 # Rationale for this change

# What changes are included in this PR?

This PR applies this for Expr::BinaryExpr.

# Are there any user-facing changes?

### CODE REVIEW

**Comment by zhoudongyan** [comment]
@andygrove @mingmwang @yjshen @jackwener PTAL, thanks

File: datafusion/core/src/physical_plan/planner.rs [hunk 1]
```diff
@@ -462,13 +462,13 @@ impl DefaultPhysicalPlanner {
...
-                    fetch,
-                    ..
-                }) => {
+                LogicalPlan::TableScan(TableScan {
+                                           source,
+                                           projection,
+                                           filters,
+                                           fetch,
+                                           ..
+                                       }) => {
```
Comment (andygrove) [inline datafusion/core/src/physical_plan/planner.rs:471]: I'm curious why there are formatting changes like this that are unrelated to making BinaryExpr a struct. Were these changes made intentionally?

File: datafusion/expr/src/expr.rs [hunk 2]
```diff
@@ -92,14 +92,7 @@ pub enum Expr {
...
     /// A binary expression such as "age > 21"
-    BinaryExpr {
-        /// Left-hand side of the expression
-        left: Box<Expr>,
-        /// The comparison operator
-        op: Operator,
-        /// Right-hand side of the expression
-        right: Box<Expr>,
-    },
+    BinaryExpr(BinaryExpr),
```
Comment (andygrove) [inline datafusion/expr/src/expr.rs:95]: :heart:

File: datafusion/optimizer/src/unwrap_cast_in_comparison.rs [hunk 3]
```diff
@@ -140,9 +138,9 @@ impl ExprRewriter for UnwrapCastExprRewriter {
                                 try_cast_literal_to_type(left_lit_value, &expr_type)?;
                             if let Some(value) = casted_scalar_value {
                                 // unwrap the cast/try_cast for the right expr
-                                return Ok(binary_expr(
+                                return Ok(datafusion_expr::binary_expr(
```
Comment (andygrove) [inline datafusion/optimizer/src/unwrap_cast_in_comparison.rs:141]: Why is the crate name required here now?

**Review by andygrove — APPROVED** [review]
LGTM. Thanks @zhoudongyan.

**Review by jackwener — APPROVED** [review]
LGTM. Except the indentation is a bit weird。

File: datafusion/core/src/physical_plan/planner.rs (hunk 1 above)
Comment (zhoudongyan) [reply datafusion/core/src/physical_plan/planner.rs:471]: They are auto-formatted by CLion IDE, let me try to to revert these unrelated changes. CC @jackwener

File: datafusion/optimizer/src/unwrap_cast_in_comparison.rs (hunk 3 above)
Comment (zhoudongyan) [reply datafusion/optimizer/src/unwrap_cast_in_comparison.rs:141]: Because we already have a local variable `binary_expr` in line 119. I can remove this crate name here and use `expr::BinaryExpr::new()`.

**Comment by zhoudongyan** [comment]
@andygrove @jackwener 

Resolved some comments
    - revert changes of unnecessary idents
    - avoid introducing `binary_expr` local variable

PTAL, many thanks. Feel free to give me some practical advice about rust because I have little experience in it.
